### PR TITLE
Updated all array_view uses with the C++20 span.

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
The updates were done in the following order
1. `AZStd::array_view<([^>].+)\* ?>`  -> `AZStd::span<\1 const>`
2. `AZStd::array_view<(?:const )(.+)>` -> `AZStd::span<const \1>`
3. `AZStd::array_view` -> `AZStd::span`

Removed the implementation of array_view from the codebase

Signed-off-by: lumberyard-employee-dm <56135373+lumberyard-employee-dm@users.noreply.github.com>

### CODE REVIEW

File: Gems/Atom/RHI/Metal/Code/Source/RHI.Builders/ShaderPlatformInterface.cpp [hunk 1]
```diff
@@ -540,28 +540,28 @@ namespace AZ
...
         }
-        
+
         bool ShaderPlatformInterface::AddConstantBufferEntries(const RHI::ShaderResourceGroupLayout& groupLayout,
                                                                AZStd::string& constantBufferTempStructs,
                                                                AZStd::string& argBufferStr,
                                                                uint32_t groupLayoutIndex) const
         {
-            AZStd::array_view<RHI::ShaderInputConstantDescriptor> shaderInputConstantList = groupLayout.GetShaderInputListForConstants();
+            AZStd::span<constRHI::ShaderInputConstantDescriptor> shaderInputConstantList = groupLayout.GetShaderInputListForConstants();
```
Comment (invertednormal) [inline Gems/Atom/RHI/Metal/Code/Source/RHI.Builders/ShaderPlatformInterface.cpp:553]: Looks like a space was removed here.

File: Gems/Atom/RHI/Metal/Code/Source/RHI/ArgumentBuffer.cpp [hunk 2]
```diff
@@ -131,8 +131,8 @@ namespace AZ
                 [argBufferDecriptors addObject:staticSamplerArgDescriptor];
                 resourceAdded = true;
             }
-            
-            AZStd::array_view<RHI::ShaderInputConstantDescriptor> shaderInputConstantList = m_srgLayout->GetShaderInputListForConstants();
+
+            AZStd::span<constRHI::ShaderInputConstantDescriptor> shaderInputConstantList = m_srgLayout->GetShaderInputListForConstants();
```
Comment (invertednormal) [inline Gems/Atom/RHI/Metal/Code/Source/RHI/ArgumentBuffer.cpp:135]: Space removed

File: Gems/Atom/RHI/Metal/Code/Source/RHI/ArgumentBuffer.cpp [hunk 3]
```diff
@@ -157,17 +157,17 @@ namespace AZ
                 [m_argumentEncoder setSamplerState:mtlSamplerState atIndex:staticSampler.m_registerId];
             }
         }
-        
+
         void ArgumentBuffer::AttachConstantBuffer()
         {
-            AZStd::array_view<RHI::ShaderInputConstantDescriptor> shaderInputConstantList = m_srgLayout->GetShaderInputListForConstants();
+            AZStd::span<constRHI::ShaderInputConstantDescriptor> shaderInputConstantList = m_srgLayout->GetShaderInputListForConstants();
```
Comment (invertednormal) [inline Gems/Atom/RHI/Metal/Code/Source/RHI/ArgumentBuffer.cpp:163]: Several more times in this file, won't add additional comments.

**Review by invertednormal — CHANGES_REQUESTED** [review]

File: Gems/Atom/RPI/Code/Tests/Shader/ShaderTests.cpp [hunk 4]
```diff
@@ -538,7 +538,7 @@ namespace UnitTest
 
             auto shaderAsset = shader->GetAsset();
             EXPECT_EQ(shader->GetPipelineStateType(), shaderAsset->GetPipelineStateType());
-            EXPECT_EQ(shader->GetShaderResourceGroupLayouts(), shaderAsset->GetShaderResourceGroupLayouts());
+            EXPECT_EQ(shader->GetShaderResourceGroupLayouts().data(), shaderAsset->GetShaderResourceGroupLayouts().data());
```
Comment (invertednormal) [inline Gems/Atom/RPI/Code/Tests/Shader/ShaderTests.cpp:541]: This is a slightly different compare than before - the ends or size need to be checked as well.

**Review by invertednormal — CHANGES_REQUESTED** [review]
Awesome - thanks for taking care of this. Just a few issues with some metal files and one compare that might need to be updated.

File: Gems/Atom/RPI/Code/Tests/Shader/ShaderTests.cpp (hunk 4 above)
Comment (lumberyard-employee-dm) [reply Gems/Atom/RPI/Code/Tests/Shader/ShaderTests.cpp:541]: I'll update it to take the size into account as well.

File: Gems/Atom/RHI/Metal/Code/Source/RHI.Builders/ShaderPlatformInterface.cpp (hunk 1 above)
Comment (lumberyard-employee-dm) [reply Gems/Atom/RHI/Metal/Code/Source/RHI.Builders/ShaderPlatformInterface.cpp]: All of these locations have been fixed.

File: Gems/Atom/RHI/Metal/Code/Source/RHI/ArgumentBuffer.h [hunk 5]
```diff
@@ -78,17 +78,17 @@ namespace AZ
...
-                                const AZStd::span<constRHI::SamplerState>& samplerStates);
+                                const AZStd::span<const RHI::SamplerState>& samplerStates);
 
             void UpdateBufferViews(const RHI::ShaderInputBufferDescriptor& shaderInputBuffer,
                                    const RHI::ShaderInputBufferIndex shaderInputIndex,
-                                   const AZStd::span<constRHI::ConstPtr<RHI::BufferView>>& bufferViews);
+                                   const AZStd::span<const RHI::ConstPtr<RHI::BufferView>>& bufferViews);
 
-            void UpdateConstantBufferViews(AZStd::span<constuint8_t> rawData);
+            void UpdateConstantBufferViews(AZStd::span<cons uint8_t> rawData);
```
Comment (invertednormal) [inline Gems/Atom/RHI/Metal/Code/Source/RHI/ArgumentBuffer.h:91]: whoops

**Review by invertednormal — CHANGES_REQUESTED** [review]
Thanks for fixing up the comments too. Almost everything fixed but it looks like the regex messed up on one thing.

File: Gems/Atom/RHI/Metal/Code/Source/RHI/ArgumentBuffer.h (hunk 5 above)
Comment (lumberyard-employee-dm) [reply Gems/Atom/RHI/Metal/Code/Source/RHI/ArgumentBuffer.h:91]: Whoops it is.

**Review by invertednormal — APPROVED** [review]

**Review by mrieggeramzn — APPROVED** [review]
Very cool change

**Pull request merged**